### PR TITLE
persist: clean up datadriven tests

### PR DESCRIPTION
**Pull request by danhhz**
- Spin out a new datadriven module with some common parsing helpers.
- Move the actual tests to the new datadriven module (for ease of selection in `cargo test`), but leave the "directive" impls as free-functions next to machine, trace (so they continue to have access to stuff that isn't `pub(crate)`).
- Standardize the HollowBatch input format the trace tests to match the output format. Previously the input looked like `0 1 2 3 k1 k2`, but now it looks like `[0][1][2] 3 k1 k2`. Paul mentioned a while ago that this was confusing, but no-one had a better idea at the time.
- A few small tweaks to the "trace" language to make things more clear between datadriven files:
  - Pass "since" as an arg to "downgrade-since" instead of an input.
  - Rename "batches" to "spine-batches".
  - Rename "insert" to "push-batch" (to match the method name).

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

I'm about to write some new datadriven tests for #14704, so wanted to clean things up a bit while I was in here.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/persist-client/src/internal/machine.rs [hunk 1]
```diff
@@ -772,356 +772,346 @@ where
...
-                                .compare_and_append(batch, &writer_id, now())
-                                .await
-                                .expect("indeterminate")
-                                .expect("invalid usage")
-                                .expect("upper mismatch");
-                            format!("ok\n")
+    /// Scans consensus and returns all states with their SeqNos
+    /// and which batches they reference
+    pub async fn consensus_scan(
+        dd: &mut MachineState,
```
Comment (danhhz) [inline src/persist-client/src/internal/machine.rs:848]: whoops missed some `dd`s

Comment (pH14) [reply src/persist-client/src/internal/machine.rs:838]: honestly just `datadriven` seems fine to me. it's verbose but very clear

File: src/persist-client/tests/trace/compaction [hunk 2]
```diff
@@ -8,16 +8,17 @@
...
-batches
+spine-batches
 ----
+<empty>
 
 # Insert a len=1 batch.
-insert
-0 1 0 1 k0
+push-batch
+[0][1][0] 1 k0
```
Comment (pH14) [inline src/persist-client/tests/trace/compaction:17]: this is so simple, but waaaaay easier to visually parse

**Review by pH14 — COMMENTED** [review]
💯 🏆 🎉 sooooooo much easier to read and reason about

**Review by pH14 — APPROVED** [review]

File: src/persist-client/src/internal/machine.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/internal/machine.rs]: done!

**Review by danhhz — COMMENTED** [review]
TFTR!

**Comment by danhhz** [comment]
One of the flakes is well known. The other is new https://github.com/MaterializeInc/materialize/issues/14861

**Pull request merged**